nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +171,194 @@ public NullnessHint onDataflowVisitMethodInvocation(
+        // Arg being non-null implies the return of the method being false.
+        elseUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // Arg being null implies the return of the method being true.
+        thenUpdates.set(accessPath, Nullness.NULL);
+      } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // Arg being non-null implies the return of the method being true.
+        thenUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // Arg being non-null implies the method throws an exception, then we can mark it as
+        // null on both non-exceptional exits from the method.
+        bothUpdates.set(accessPath, Nullness.NULL);
+      } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // Arg being null implies the method throws an exception, then we can mark it as
+        // non-null on both non-exceptional exits from the method.
+        bothUpdates.set(accessPath, Nullness.NONNULL);
       }
     }
     return NullnessHint.UNKNOWN;
   }
+
+  private static final class IsNullLiteralNodeVisitor extends AbstractNodeVisitor<Boolean, None> {
+    private static final IsNullLiteralNodeVisitor INSTANCE = new IsNullLiteralNodeVisitor();
+
+    @Override
+    public Boolean visitNode(Node node, None unused) {
+      return false;
+    }
+
+    @Override
+    public Boolean visitNullLiteral(NullLiteralNode node, None unused) {
+      return true;
+    }
+  }
+
+  private static final class NullEqualityVisitor extends AbstractNodeVisitor<Optional<Node>, None> {
+
+    private static final NullEqualityVisitor IS_NULL = new NullEqualityVisitor(true);
+    private static final NullEqualityVisitor NOT_NULL = new NullEqualityVisitor(false);
+
+    private final boolean equals;
+
+    NullEqualityVisitor(boolean equals) {
+      this.equals = equals;
+    }
+
+    @Override
+    public Optional<Node> visitNode(Node node, None unused) {
+      return Optional.empty();
+    }
+
+    @Override
+    public Optional<Node> visitNotEqual(NotEqualNode notEqualNode, None unused) {
+      if (equals) {
+        return Optional.empty();
+      } else {
+        return visit(notEqualNode);
+      }
+    }
+
+    @Override
+    public Optional<Node> visitEqualTo(EqualToNode equalToNode, None unused) {
+      if (equals) {
+        return visit(equalToNode);
+      } else {
+        return Optional.empty();
+      }
+    }
+
+    private Optional<Node> visit(BinaryOperationNode comparison) {
+      Node lhs = comparison.getLeftOperand();
+      Node rhs = comparison.getRightOperand();
+      boolean lhsIsNullLiteral = lhs.accept(IsNullLiteralNodeVisitor.INSTANCE, None.INSTANCE);
+      boolean rhsIsNullLiteral = rhs.accept(IsNullLiteralNodeVisitor.INSTANCE, None.INSTANCE);
```

Comment:
We can do better here by matching expressions where one side evaluates to `Nullness.NULL` and the other `Nullness.NULLABLE`, allowing for slightly more complex inputs.

This PR has become broader than I'd prefer, so I'll bias toward investigating such improvements in a follow-up PR unless it's requested in this one.